decavision/dataset_preparation/generate_pseudolabels.py

```diff
@@ -0,0 +1,216 @@
+            for i in raw_predictions_all:
+                predictions.append(i[idx])
+
+            title = "Confidences for the class: {}".format(classes[idx])
+            path = "outputs/{}_confidences.png".format(classes[idx])
+            self._plot_data(predictions, title, path)
+
+        # Plot graph for highest confidence pseudo labels for all unlabeled images
+        self._plot_data(raw_predictions,
+                        name="Highest confidence pseudo labels",
+                        output_path="outputs/highest_confidence_predictions.png")
+
+    def _save_pseudo_labels(self, csv_path, threshold, dictionary):
+        """
+        Save pseudo labels.
+
+        Arguments:
+            csv_path (str): Path to CSV file
+            threshold (float): Discard images with prediction below this confidence, default is None.
+            dictionary (dict): Class dictionary
+        """
+
+        dt = pd.read_csv(csv_path)
+        filepaths = dt[["Filepaths"]].values
+        predicted_class = dt[["Predicted Class"]].values
+        raw_predictions = dt[["Highest Confidence"]].values
+
+        for proba, y, path in zip(raw_predictions, predicted_class, filepaths):
+            # The results variable should be the same as the class category
+            for class_name, index in dictionary.items():
+                if threshold:
+                    # For thresholding predictions
+                    if index == y and proba >= threshold:
+                        shutil.copy(os.path.join(self.unlabeled_path, str(path[0])),
+                                    os.path.join(self.pseudo_data_path, class_name))
+                else:
+                    # For hard predictions
+                    if index == y:
+                        shutil.copy(os.path.join(self.unlabeled_path, str(path[0])),
+                                    os.path.join(self.pseudo_data_path, class_name))
+        print("Saved pseudo labels.")
+
+    def generate_pseudolabel_data(self, save_confidences_csv=False,
+                                  plot_confidences=False, threshold=None):
+        """ Convert image and label to tfrecord example.
+            Arguments:
+                save_confidences_csv (boolean): Whether to save CSV file with pseudo confidences, default is False.
+                plot_confidences (boolean): Whether to plot confidence graphs for raw confidences and per class confidences.
+                threshold (float): Discard images with prediction below this confidence, default is None.
+
+            Returns:
+                pseudo_data_path: A folder with both labeled and pseudo labeled images.
+        """
+
+        # Copy the training/labeled data to the destination folder where
+        # we will also store the pseudo labels.
+        copy_tree(self.train_data_path, self.pseudo_data_path)
+
+        # Make dictionary for classes and their index
+        class_names = sorted(os.listdir(self.train_data_path))
+        class_dict = {}
```

Comment:
Here are two suggestions to make this code much cleaner:
- use enumerate instead of making your own index.
- use list/dict comprehension.

The result will look like
```
class_dict = {cat, i for i, cat in enumerate(class_names)}
```

Reply:
Done.